<launch>
  <!-- ==================================== -->
  <!-- Unitree Go1 SSC-based Navigation System -->
  <!-- ==================================== -->

  <!-- Arguments -->
  <arg name="robot_type" default="go1" />
  <arg name="use_sim" default="true" />
  <arg name="use_dsfnet" default="true" />
  <arg name="dsfnet_model_path" default="" />

  <!-- Map size for ego-planner -->
  <arg name="map_size_x" value="20.0"/>
  <arg name="map_size_y" value="20.0"/>
  <arg name="map_size_z" value="2.5"/>

  <!-- Odometry topic -->
  <arg name="odom_topic" value="/visual_slam/odom" />

  <!-- Camera topics -->
  <arg name="rgb_topic" default="/camera/color/image_raw" />
  <arg name="depth_topic" default="/camera/depth/image_raw" />
  <arg name="camera_info_topic" default="/camera/color/camera_info" />

  <!-- ============================================ -->
  <!-- 1. Unitree Robot (Simulation or Real) -->
  <!-- ============================================ -->
  <group if="$(arg use_sim)">
    <include file="$(find unitree_guide)/launch/gazeboSim.launch">
      <arg name="rname" value="$(arg robot_type)" />
    </include>

    <!-- Start Gazebo odometry broadcaster -->
    <node pkg="gazebo_odom_broadcaster" type="gazebo_odom_broadcaster_node" name="gazebo_odom_broadcaster" output="screen"/>
  </group>

  <!-- For real robot, uncomment below -->
  <!-- <group unless="$(arg use_sim)">
    <include file="$(find unitree_legged_real)/launch/real.launch">
      <arg name="ctrl_level" value="highlevel"/>
    </include>
  </group> -->

  <!-- ============================================ -->
  <!-- 2. Odometry Bridge -->
  <!-- ============================================ -->
  <node pkg="unitree_ssc_navigation" type="unitree_odom_bridge.py" name="unitree_odom_bridge" output="screen">
    <param name="odom_input_topic" value="/trunk_odom" />
    <param name="odom_output_topic" value="$(arg odom_topic)" />
    <param name="pose_output_topic" value="/mavros/local_position/pose" />
    <param name="camera_frame" value="camera_color_optical_frame" />
    <param name="base_frame" value="base" />
    <param name="world_frame" value="world" />
    <param name="camera_x" value="0.28" />
    <param name="camera_y" value="0.0" />
    <param name="camera_z" value="0.03" />
  </node>

  <!-- ============================================ -->
  <!-- 3. DSFNet Semantic Scene Completion (Optional) -->
  <!-- ============================================ -->
  <group if="$(arg use_dsfnet)">
    <node pkg="unitree_ssc_navigation" type="dsfnet_inference_node.py" name="dsfnet_inference" output="screen">
      <param name="model_path" value="$(arg dsfnet_model_path)" />
      <param name="num_classes" value="12" />
      <param name="fx" value="387.229248046875" />
      <param name="fy" value="387.229248046875" />
      <param name="cx" value="321.04638671875" />
      <param name="cy" value="243.44969177246094" />
      <param name="img_height" value="480" />
      <param name="img_width" value="640" />

      <!-- Remap camera topics -->
      <remap from="/camera/color/image_raw" to="$(arg rgb_topic)" />
      <remap from="/camera/depth/image_raw" to="$(arg depth_topic)" />
    </node>

    <!-- SSC to Planner Bridge -->
    <node pkg="unitree_ssc_navigation" type="ssc_to_planner_bridge.py" name="ssc_to_planner_bridge" output="screen">
      <param name="ground_height_threshold" value="0.3" />
      <param name="max_height" value="2.0" />
      <param name="min_height" value="-0.5" />
      <param name="max_range" value="10.0" />
    </node>
  </group>

  <!-- ============================================ -->
  <!-- 4. Ego-Planner -->
  <!-- ============================================ -->
  <include file="$(find ego_planner)/launch/advanced_param.xml">
    <arg name="map_size_x_" value="$(arg map_size_x)"/>
    <arg name="map_size_y_" value="$(arg map_size_y)"/>
    <arg name="map_size_z_" value="$(arg map_size_z)"/>
    <arg name="odometry_topic" value="$(arg odom_topic)"/>

    <!-- Use SSC point cloud if enabled, otherwise use direct depth -->
    <arg name="cloud_topic" value="/pcl_render_node/cloud" unless="$(arg use_dsfnet)"/>
    <arg name="cloud_topic" value="/ssc/obstacle_cloud" if="$(arg use_dsfnet)"/>

    <!-- Camera intrinsics -->
    <arg name="cx" value="321.04638671875"/>
    <arg name="cy" value="243.44969177246094"/>
    <arg name="fx" value="387.229248046875"/>
    <arg name="fy" value="387.229248046875"/>

    <!-- Ground robot parameters (lower speed than drone) -->
    <arg name="max_vel" value="0.8" />
    <arg name="max_acc" value="1.5" />

    <!-- Planning horizon -->
    <arg name="planning_horizon" value="5.0" />

    <!-- Use 2D Nav Goal -->
    <arg name="flight_type" value="1" />
  </include>

  <!-- ============================================ -->
  <!-- 5. Trajectory Server (MPC Controller) -->
  <!-- ============================================ -->
  <node pkg="ego_planner" name="traj_server" type="traj_server" output="screen">
    <remap from="/position_cmd" to="planning/pos_cmd"/>
    <remap from="/odom_world" to="$(arg odom_topic)"/>
    <param name="traj_server/time_forward" value="1.0" type="double"/>
  </node>

  <!-- ============================================ -->
  <!-- 6. Waypoint Generator -->
  <!-- ============================================ -->
  <node pkg="waypoint_generator" name="waypoint_generator" type="waypoint_generator" output="screen">
    <remap from="~odom" to="$(arg odom_topic)"/>
    <remap from="~goal" to="/move_base_simple/goal"/>
    <remap from="~traj_start_trigger" to="/traj_start_trigger" />
    <param name="waypoint_type" value="manual-lonely-waypoint"/>
  </node>

  <!-- ============================================ -->
  <!-- 7. Unitree Controller -->
  <!-- ============================================ -->
  <!-- The unitree_guide controller should be started separately using: -->
  <!-- rosrun unitree_guide junior_ctrl -->
  <!-- It will automatically subscribe to /cmd_vel from traj_server -->

  <!-- ============================================ -->
  <!-- 8. Visualization -->
  <!-- ============================================ -->
  <node type="rviz" name="rviz" pkg="rviz" args="-d $(find unitree_ssc_navigation)/config/navigation.rviz" if="$(arg use_sim)"/>

</launch>
